mllib/src/test/scala/org/apache/spark/ml/util/MLTestingUtils.scala

```diff
@@ -224,4 +208,139 @@ object MLTestingUtils extends SparkFunSuite {
+    require(noiseLevel >= 0.0 && noiseLevel <= 1.0, "noiseLevel must be in range [0.0, 1.0]")
+    val rng = new scala.util.Random(seed)
+    val predictionFunc = getRandomLinearPredictionFunction(numFeatures, numClasses, seed)
+    Range(0, numPoints).map { i =>
+      val features = Vectors.dense(Array.tabulate(numFeatures) { j =>
+        val numCategories = categoricalFeaturesInfo.getOrElse(j, 0)
+        if (numCategories > 0) {
+          rng.nextInt(numCategories)
+        } else {
+          rng.nextDouble() - 0.5
+        }
+      })
+      val label = predictionFunc(features)
+      val noisyLabel = if (numClasses > 0) {
+        // with probability equal to noiseLevel, select a random class instead of the true class
+        if (rng.nextDouble < noiseLevel) rng.nextInt(numClasses) else label
+      } else {
+        // add noise to the label proportional to the noise level
+        label + noiseLevel * rng.nextGaussian()
+      }
+      Instance(noisyLabel, 1.0, features)
+    }
+  }
+
+  /**
+   * Helper function for testing sample weights. Tests that oversampling each point is equivalent
+   * to assigning a sample weight proportional to the number of samples for each point.
+   */
+  def testOversamplingVsWeighting[M <: Model[M], E <: Estimator[M]](
+        spark: SparkSession,
+        estimator: E with HasWeightCol with HasLabelCol with HasFeaturesCol,
+        categoricalFeaturesInfo: Map[Int, Int],
+        numPoints: Int,
+        numClasses: Int,
+        numFeatures: Int,
+        modelEquals: (M, M) => Unit,
+        seed: Long): Unit = {
+    import spark.implicits._
+    val df = generateNoisyData(numPoints, numClasses, numFeatures, categoricalFeaturesInfo,
+      seed).toDF()
+    val (overSampledData, weightedData) = genEquivalentOversampledAndWeightedInstances(
+      df, estimator.getLabelCol, estimator.getFeaturesCol, seed)
+    val weightedModel = estimator.set(estimator.weightCol, "weight").fit(weightedData)
+    val overSampledModel = estimator.set(estimator.weightCol, "").fit(overSampledData)
+    modelEquals(weightedModel, overSampledModel)
+  }
+
+  /**
+   * Helper function for testing sample weights. Tests that injecting a large number of outliers
+   * with very small sample weights does not affect fitting. The predictor should learn the true
+   * model despite the outliers.
+   */
+  def testOutliersWithSmallWeights[M <: Model[M], E <: Estimator[M]](
+        spark: SparkSession,
+        estimator: E with HasWeightCol with HasLabelCol with HasFeaturesCol,
+        categoricalFeaturesInfo: Map[Int, Int],
+        numPoints: Int,
+        numClasses: Int,
+        numFeatures: Int,
+        modelEquals: (M, M) => Unit,
+        seed: Long): Unit = {
+    import spark.implicits._
+    val df = generateNoisyData(numPoints, numClasses, numFeatures, categoricalFeaturesInfo,
+      seed).toDF()
+    val outlierFunction = getRandomLinearPredictionFunction(numFeatures, numClasses, seed - 1)
```

Comment:
I'm more prefer to implement ```outlierFunction``` as a simple function such as:
* ```class -> numClass - class - 1``` for classification.
* ```label -> -label``` for regression.

which should be more intuitional and easy to be understand by developers/contributors.